File: API/models/wakaf-abadi.js
'use strict';
const moment = require('moment-timezone');
const {
    Model
} = require('sequelize');
module.exports = (sequelize, DataTypes) => {
    class WakafAbadi extends Model {
        /**
         * Helper method for defining associations.
         * This method is not a part of Sequelize lifecycle.
         * The `models/index` file will call this method automatically.
         */
        static associate(models) {
            WakafAbadi.belongsTo(models.ProgramWakaf, {
                foreignKey: 'program_wakaf_id',
                targetKey: 'id',
                as: 'program_wakaf',
            });
            WakafAbadi.belongsTo(models.Wakif, {
                foreignKey: 'wakif_id',
                targetKey: 'id',
                as: 'wakif',
            });
        }
    };
    WakafAbadi.init({
            id: {
                type: DataTypes.UUID,
                allowNull: false,
                primaryKey: true,
            },
            program_wakaf_id: {
                type: DataTypes.INTEGER,
                allowNull: false,
                references: {
                    model: 'program_wakaf',
                    key: 'id',
                },
            },
            wakif_id: {
                type: DataTypes.INTEGER,
                allowNull: false,
                references: {
                    model: 'wakif',
                    key: 'id',
                },
            },
            nominal: {
                type: DataTypes.INTEGER,
                allowNull: false,
            },
            nama_wakif: {
                type: DataTypes.STRING,
                allowNull: false,
            },
            metode_pembayaran: {
                type: DataTypes.STRING,
                allowNull: false,
            },
            kode_pembayaran: {
                type: DataTypes.TEXT,
                allowNull: true,
            },
            status_pembayaran: {
                type: DataTypes.ENUM,
                allowNull: false,
                values: ['capture', 'settlement', 'pending', 'deny', 'cancel', 'expire', 'refund'],
                defaultValue: 'pending',
            },
            nama_file_sertifikat: {
                type: DataTypes.STRING,
                allowNull: true,
            },
            nama_file_akta: {
                type: DataTypes.STRING,
                allowNull: true,
            },
            created_at: {
                type: DataTypes.DATE,
                allowNull: false,
            },
            updated_at: {
                type: DataTypes.DATE,
                allowNull: false,
            },
            tanggal: {
                type: DataTypes.VIRTUAL,
                get() {
                    return moment(this.getDataValue('created_at')).locale('id').tz('Asia/Jakarta').format('dddd, LL');
                },
                set(value) {
                    throw new Error('Do not try to set the value!');
                }
            },
            waktu: {
                type: DataTypes.VIRTUAL,
                get() {
                    return moment(this.getDataValue('created_at')).locale('id').tz('Asia/Jakarta').format('HH:mm');
                },
                set(value) {
                    throw new Error('Do not try to set the value!');
                }
            },
            jenis: {
                type: DataTypes.VIRTUAL,
                get() {
                    return "Abadi";
                },
                set(value) {
                    throw new Error('Do not try to set the value!');
                }
            }
        },
        {
            sequelize,
            modelName: 'WakafAbadi',
            tableName: 'wakaf_abadi',
            createdAt: 'created_at',
            updatedAt: 'updated_at',
            deletedAt: false,
            underscored: true,
        });
    return WakafAbadi;
};

File: Mobile/lib/model/wakaf_abadi.dart
import 'package:flutter_cash_waqf/model/program_wakaf.dart';
import 'package:flutter_cash_waqf/model/wakaf.dart';
import 'package:flutter_cash_waqf/utilities/constant.dart';

class WakafAbadi extends Wakaf {
  String? id;
  int? programWakafId;
  int? wakifId;
  int? nominal;
  String? namaWakif;
  String? jenisPembayaran;
  String? metodePembayaran;
  String? kodePembayaran;
  String? statusPembayaran;
  String? aktaIkrarWakaf;
  String? sertifikatWakafUang;
  String? tanggal;
  String? waktu;
  ProgramWakaf? programWakaf;

  WakafAbadi({
    this.id,
    this.programWakafId,
    this.wakifId,
    this.nominal,
    this.namaWakif,
    this.jenisPembayaran,
    this.metodePembayaran,
    this.kodePembayaran,
    this.statusPembayaran,
    this.aktaIkrarWakaf,
    this.sertifikatWakafUang,
    this.tanggal,
    this.waktu,
    this.programWakaf,
  });

  factory WakafAbadi.createFromJson(Map<String, dynamic> data) => WakafAbadi(
        id: data['id'],
        nominal: data['nominal'],
        namaWakif: data['nama_wakif'],
        metodePembayaran: data['metode_pembayaran'],
        kodePembayaran: data['kode_pembayaran'],
        statusPembayaran: data['status_pembayaran'],
        aktaIkrarWakaf: (data['nama_file_akta'] != null)
            ? "${Constant.aiwUrlApi}/${data['nama_file_akta']}"
            : null,
        sertifikatWakafUang: (data['nama_file_sertifikat'] != null)
            ? "${Constant.swuUrlApi}/${data['nama_file_sertifikat']}"
            : null,
        tanggal: data['tanggal'],
        waktu: data['waktu'],
        programWakaf: ProgramWakaf.createWithoutKategori(data['program_wakaf']),
      );

  factory WakafAbadi.createWithoutProgram(Map<String, dynamic> data) =>
      WakafAbadi(
        id: data['id'],
        nominal: data['nominal'],
        namaWakif: data['nama_wakif'],
        metodePembayaran: data['metode_pembayaran'],
        kodePembayaran: data['kode_pembayaran'],
        statusPembayaran: data['status_pembayaran'],
        tanggal: data['tanggal'],
        waktu: data['waktu'],
      );
}
